File: only-one-page-rule-applies/python/cloudflare_page_rule_shadow.py
"""Find Page Rules that match but never run.

Only the highest-priority matching rule takes effect. Every other match is
discarded silently, so a rule can be present, enabled, correct and dead.
"""
import argparse
import fnmatch
import logging
import os
import sys
from urllib.parse import urlsplit

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("cloudflare_page_rule_shadow")

API = "https://api.cloudflare.com/client/v4"


def normalise(pattern, url):
    """Strip the optional scheme from both sides before comparing.

    A pattern with no scheme matches http and https. Comparing raw strings would
    make such a pattern look narrower than it is -- the opposite of the truth.
    """
    if "://" not in pattern:
        url = urlsplit(url).netloc + urlsplit(url).path + (
            "?" + urlsplit(url).query if urlsplit(url).query else "")
    return pattern, url


def matches(pattern, url):
    pat, target = normalise(pattern, url)
    return fnmatch.fnmatch(target, pat)


def evaluate(rules, url):
    """Return (winner, shadowed) for one URL.

    rules: list of dicts with 'priority', 'pattern', 'enabled', 'actions'.
    Higher priority wins. Disabled rules never match at all.
    """
    active = sorted((r for r in rules if r.get("enabled", True)),
                    key=lambda r: -r["priority"])
    hits = [r for r in active if matches(r["pattern"], url)]
    if not hits:
        return None, []
    return hits[0], hits[1:]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--zone-id", required=True)
    ap.add_argument("--url", nargs="+", required=True)
    args = ap.parse_args()

    token = os.environ.get("CF_API_TOKEN")
    if not token:
        log.error("set CF_API_TOKEN")
        return 2

    r = requests.get(f"{API}/zones/{args.zone_id}/pagerules",
                     headers={"Authorization": f"Bearer {token}"}, timeout=30)
    r.raise_for_status()
    raw = r.json().get("result", [])

    rules = [{
        "priority": item.get("priority", 0),
        "pattern": item["targets"][0]["constraint"]["value"],
        "enabled": item.get("status") == "active",
        "actions": [a.get("id") for a in item.get("actions", [])],
    } for item in raw]

    disabled = [x for x in rules if not x["enabled"]]
    for d in disabled:
        log.warning("DISABLED  %s -- still counts against your rule quota", d["pattern"])

    shadowed_any = False
    for url in args.url:
        winner, shadowed = evaluate(rules, url)
        if not winner:
            log.info("%s -- no Page Rule matches", url)
            continue
        log.info("%s -> %s  actions=%s", url, winner["pattern"], winner["actions"])
        for s in shadowed:
            shadowed_any = True
            log.error("  SHADOWED  %s (actions=%s) matches but never runs -- only the "
                      "highest-priority match applies", s["pattern"], s["actions"])
    return 1 if shadowed_any else 0


if __name__ == "__main__":
    sys.exit(main())

File: only-one-page-rule-applies/node/cloudflare-page-rule-shadow.mjs
/**
 * Find Page Rules that match but never run.
 *
 * Only the highest-priority matching rule takes effect. Every other match is
 * discarded silently, so a rule can be present, enabled, correct and dead.
 */
const API = 'https://api.cloudflare.com/client/v4';

const toRegExp = (pattern) => new RegExp(
  `^${pattern.replace(/[.+^${}()|[\]\\]/g, '\\$&').replace(/\*/g, '.*')}$`);

/**
 * Strip the optional scheme from both sides before comparing. A pattern with no
 * scheme matches http and https, so a raw string compare makes it look narrower
 * than it is -- the opposite of the truth.
 */
export function matches(pattern, url) {
  const target = pattern.includes('://') ? url : url.replace(/^https?:\/\//, '');
  return toRegExp(pattern).test(target);
}

/** Return { winner, shadowed } for one URL. Disabled rules never match at all. */
export function evaluate(rules, url) {
  const active = rules.filter((r) => r.enabled !== false).sort((a, b) => b.priority - a.priority);
  const hits = active.filter((r) => matches(r.pattern, url));
  return { winner: hits[0] ?? null, shadowed: hits.slice(1) };
}

async function main() {
  const zone = process.argv[process.argv.indexOf('--zone-id') + 1];
  const ui = process.argv.indexOf('--url');
  const urls = process.argv.slice(ui + 1).filter((a) => !a.startsWith('--'));
  const token = (process.env.CF_API_TOKEN || "");
  if (!token) { console.error('set CF_API_TOKEN'); process.exit(2); }

  const res = await fetch(`${API}/zones/${zone}/pagerules`,
    { headers: { Authorization: `Bearer ${token}` } });
  const { result: raw = [] } = await res.json();

  const rules = raw.map((item) => ({
    priority: item.priority ?? 0,
    pattern: item.targets[0].constraint.value,
    enabled: item.status === 'active',
    actions: (item.actions ?? []).map((a) => a.id),
  }));

  for (const d of rules.filter((r) => !r.enabled)) {
    console.warn(`DISABLED  ${d.pattern} -- still counts against your rule quota`);
  }

  let shadowedAny = false;
  for (const url of urls) {
    const { winner, shadowed } = evaluate(rules, url);
    if (!winner) { console.log(`${url} -- no Page Rule matches`); continue; }
    console.log(`${url} -> ${winner.pattern}  actions=${winner.actions}`);
    for (const s of shadowed) {
      shadowedAny = true;
      console.error(`  SHADOWED  ${s.pattern} (actions=${s.actions}) matches but never runs`);
    }
  }
  process.exit(shadowedAny ? 1 : 0);
}

if (import.meta.url === `file://${process.argv[1]}`) main();
